cpp-client/deephaven/dhclient/src/impl/client_impl.cc

```diff
@@ -25,25 +24,55 @@ std::shared_ptr<ClientImpl> ClientImpl::Create(
     std::shared_ptr<Executor> executor,
     std::shared_ptr<Executor> flight_executor,
     const std::string &session_type) {
-  std::optional<Ticket> consoleTicket;
+  std::optional<Ticket> console_ticket;
   if (!session_type.empty()) {
     auto cb = SFCallback<StartConsoleResponse>::CreateForFuture();
     server->StartConsoleAsync(session_type, std::move(cb.first));
     StartConsoleResponse scr = std::move(std::get<0>(cb.second.get()));
-    consoleTicket = std::move(*scr.mutable_result_id());
+    console_ticket = std::move(*scr.mutable_result_id());
   }
 
   auto thmi = TableHandleManagerImpl::Create(
-          std::move(consoleTicket),
+          std::move(console_ticket),
           std::move(server),
           std::move(executor),
           std::move(flight_executor));
   return std::make_shared<ClientImpl>(Private(), std::move(thmi));
 }
 
 ClientImpl::ClientImpl(Private, std::shared_ptr<TableHandleManagerImpl> &&manager_impl) :
-    managerImpl_(std::move(manager_impl)) {}
+    manager_impl_(std::move(manager_impl)) {}
 
 ClientImpl::~ClientImpl() = default;
+
+ClientImpl::OnCloseCbId ClientImpl::AddOnCloseCallback(OnCloseCb cb) {
+  std::unique_lock lock(on_close_mux_);
+  if (on_close_cb_ctx_ == nullptr) {
+    on_close_cb_ctx_ = std::make_unique<OnCloseCbContext>();
+  }
+  OnCloseCbId id({on_close_cb_ctx_->next_id_++});
+  on_close_cb_ctx_->map_[id] = std::move(cb);
+  return id;
+}
+
+bool ClientImpl::RemoveOnCloseCallback(OnCloseCbId cb_id) {
+  std::unique_lock lock(on_close_mux_);
+  if (on_close_cb_ctx_ == nullptr) {
+    return false;
+  }
+  return on_close_cb_ctx_->map_.erase(std::move(cb_id)) > 0;
+}
+
+void ClientImpl::Shutdown() {
+  std::unique_lock lock(on_close_mux_);
+  manager_impl_->Shutdown();
+  if (on_close_cb_ctx_ == nullptr) {
+    return;
+  }
+  for (const auto &entry : on_close_cb_ctx_->map_) {
+    entry.second();
+  }
+  on_close_cb_ctx_.reset();
+}
```

Comment:
Holding a lock while invoking a user callback is kind of an antipattern. I prefer a style where you make the collection "local" and then you work through it. (I also don't see why we're calling `manager_impl_->Shutdown` under lock, as we have no protection against calling it twice anyway)

I'd recommend something like
```
void ClientImpl::Shutdown() {
  manager_impl_->Shutdown();

  on_close_mux_.lock();
  auto local_ctx = std::move(on_close_cb_ctx_);
  on_close_mux_.unlock();

  if (local_ctx == nullptr) {
    return;
  }
  for (const auto &entry : local_ctx->map_) {
    entry.second();
  }
}
```

Or if we want to protect against multiple calls to `Shutdown`, add a boolean flag, tested and set under lock.